tensorflow/core/grappler/optimizers/arithmetic_optimizer.cc

```diff
@@ -1871,6 +1871,102 @@ class FuseSquaredDiffStage : public ArithmeticOptimizerStage {
+class FuseMulAddStage : public ArithmeticOptimizerStage {
+ public:
+  explicit FuseMulAddStage(const GraphOptimizerContext& ctx,
+                                const ArithmeticOptimizerContext& ctx_ext)
+      : ArithmeticOptimizerStage("FuseMulAddStage", ctx, ctx_ext) {}
+  ~FuseMulAddStage() override = default;
+
+  bool IsSupported(const NodeDef* node) const override {
+    if(!IsAdd(*node) && !IsSub(*node))
+      return false; 
+    // todo: handle rank>5
+    auto dtype = GetDataTypeFromAttr(*node, "T");
+    return (dtype==DT_HALF || dtype==DT_FLOAT || dtype==DT_DOUBLE);
+  }
+
+  Status TrySimplify(NodeDef* node, string* simplified_node_name) override {
+    NodeDef* b, *c;
+    TF_RETURN_IF_ERROR(GetInputNode(node->input(0), &b));
+    TF_RETURN_IF_ERROR(GetInputNode(node->input(1), &c));
+    auto dtype = GetDataTypeFromAttr(*node, "T");
+    bool add = IsAdd(*node);
+    bool can_absorb_b = (IsMul(*b) && !IsInPreserveSet(*b) &&
+        (NumNonControlOutputs(*b, *ctx().node_map) == 1));
+    bool can_absorb_c = (IsMul(*c) && !IsInPreserveSet(*c) &&
+        (NumNonControlOutputs(*c, *ctx().node_map) == 1));
+    if(GetDataTypeFromAttr(*b, "T") != dtype
+      ||GetDataTypeFromAttr(*c, "T") != dtype) {
+      VLOG(1)<<"Input dtype mismatch"; // apparently this can happen
+      return Status::OK();
+    }
+
+    if(can_absorb_b) {
+      NodeDef *b0, *b1;
+      TF_RETURN_IF_ERROR(GetInputNode(b->input(0), &b0));
+      TF_RETURN_IF_ERROR(GetInputNode(b->input(1), &b1));
+      auto dtype_b0 = GetDataTypeFromAttr(*b0, "T");
+      auto dtype_b1 = GetDataTypeFromAttr(*b1, "T");
+      if(dtype_b0!=dtype || dtype_b1!=dtype) {
+        VLOG(1)<<"Input 1 inputs dtype mismatch";
+        can_absorb_b = false;
+      }
+    }
+    if(can_absorb_c) {
+      NodeDef *b0, *b1;
+      TF_RETURN_IF_ERROR(GetInputNode(c->input(0), &b0));
+      TF_RETURN_IF_ERROR(GetInputNode(c->input(1), &b1));
+      auto dtype_b0 = GetDataTypeFromAttr(*b0, "T");
+      auto dtype_b1 = GetDataTypeFromAttr(*b1, "T");
+      if(dtype_b0!=dtype || dtype_b1!=dtype) {
+        VLOG(1)<<"Input 2 inputs dtype mismatch";
+        can_absorb_c = false;
+      }
+    }
+    VLOG(1)<<"FuseMulAddSimplify "<<node->name()<<" "
+        <<node->input(0)<<" "<<node->input(1)<<" "<<can_absorb_b<<" "<<can_absorb_c<<"\n"<<node->DebugString()<<"\n"<<b->DebugString()<<"\n"<<c->DebugString();
+    if(can_absorb_b && can_absorb_c) {
+      node->set_op(add ? "FusedMulAdd2" : "FusedMulSub2");
+      node->add_input(node->input(1));
+      node->add_input(c->input(1));
+      node->set_input(1, b->input(1));
+      b->set_op("Identity");
+      c->set_op("Identity");
+      AddToOptimizationQueue(node);
+      AddToOptimizationQueue(b);
+      AddToOptimizationQueue(c);
+    }
+    else if(can_absorb_b) {
+      node->set_op(add ? "FusedMulAdd" : "FusedMulSub");
+      node->add_input(node->input(1));
+      node->set_input(1, b->input(1));
+      // 0 stays b
```

Comment:
this threw me for a loop, but then I realized this works because "b" is being changed to an "Identity" op.  A comment here stating just that would be helpful.

Also there seems to be an assumption here that the second input to "b" will be the one that gets dropped...do we need to explicitly drop it?

Is it possible to drop "b" altogether and hook up both its inputs directly to the FMA op?

Reply:
I am copying the pattern used 50 lines higher in FuseSquaredDiffStage. I am not sure if it is possible to drop the nodes. It would simplify the code, but it's unnecessary since the optimizer will, presumably, drop the identity ops on its own